templates/bootstrap.sh

```diff
@@ -9,106 +29,99 @@ function buildbox-prompt-and-run {
-    BUILDBOX_BIN_PATH="$BUILDBOX_PATH/bin"
-  else
-    BUILDBOX_BIN_PATH="$BUILDBOX_PATH"
-  fi
-fi
+: ${BUILDBOX_PATH:="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"}
+: ${BUILDBOX_BIN_PATH:="$BUILDBOX_PATH/bin"}
+: ${BUILDBOX_BUILD_PATH:="$BUILDBOX_PATH/builds"}
 
 # Add the $BUILDBOX_BIN to the $PATH
 export PATH="$BUILDBOX_BIN_PATH:$PATH"
 
-# Create the build directory
+# Send the bootstrap version back to Buildbox
+buildbox-agent build-data set "buildkite:bootstrap:version" $BUILDBOX_BOOTSTRAP_VERSION
+
+# Come up with the place that the repository will be checked out to
 SANITIZED_AGENT_NAME=$(echo $BUILDBOX_AGENT_NAME | tr -d '"')
 PROJECT_FOLDER_NAME="$SANITIZED_AGENT_NAME/$BUILDBOX_PROJECT_SLUG"
+BUILDBOX_BUILD_CHECKOUT_PATH="$BUILDBOX_BUILD_PATH/$PROJECT_FOLDER_NAME"
 
-# Allow overiding the location that builds get run in
-if [ -z "$BUILDBOX_BUILD_PATH" ]; then
-  BUILDBOX_PROJECT_PATH="$BUILDBOX_PATH/builds/$PROJECT_FOLDER_NAME"
-else
-  BUILDBOX_PROJECT_PATH="$BUILDBOX_BUILD_PATH/$PROJECT_FOLDER_NAME"
-fi
+##############################################################
+#
+# REPOSITORY HANDLING
+# Creates the build folder and makes sure we're running the
+# build at the right commit.
+#
+##############################################################
 
-# Show the ENV variables if DEBUG is on
-if [ "$BUILDBOX_AGENT_DEBUG" == "true" ]
-then
-  buildbox-run "env | grep BUILDBOX"
+# Remove the checkout folder if BUILDBOX_CLEAN_CHECKOUT is present
+if [[ "$BUILDBOX_CLEAN_CHECKOUT" == "true" ]]; then
+  echo '--- Cleaning project checkout'
+
+  buildbox-run "rm -rf \"$BUILDBOX_BUILD_CHECKOUT_PATH\""
 fi
 
-buildbox-run "mkdir -p \"$BUILDBOX_PROJECT_PATH\""
-buildbox-run "cd \"$BUILDBOX_PROJECT_PATH\""
+echo '--- Preparing build folder'
+
+buildbox-run "mkdir -p \"$BUILDBOX_BUILD_CHECKOUT_PATH\""
+buildbox-run "cd \"$BUILDBOX_BUILD_CHECKOUT_PATH\""
 
 # Do we need to do a git checkout?
-if [ ! -d ".git" ]
-then
+if [[ ! -d ".git" ]]; then
   # If it's a first time SSH git clone it will prompt to accept the host's
   # fingerprint. To avoid this add the host's key to ~/.ssh/known_hosts ahead
   # of time:
   #   ssh-keyscan -H host.com >> ~/.ssh/known_hosts
-  buildbox-run "git clone "$BUILDBOX_REPO" . -qv"
-fi
-
-# Default empty branch names
-if [ "$BUILDBOX_BRANCH" == "" ]
-then
-  BUILDBOX_BRANCH="master"
+  buildbox-run "git clone \"$BUILDBOX_REPO\" . -qv"
 fi
 
 buildbox-run "git clean -fdq"
 buildbox-run "git fetch -q"
 
-# Only reset to the branch if we're not on a tag
-if [ "$BUILDBOX_TAG" == "" ]
-then
-buildbox-run "git reset --hard origin/$BUILDBOX_BRANCH"
+# Allow checkouts of forked pull requests on GitHub only. See:
+# https://help.github.com/articles/checking-out-pull-requests-locally/#modifying-an-inactive-pull-request-locally
+if [[ "$BUILDBOX_PULL_REQUEST" != "false" ]] && [[ "$BUILDBOX_PROJECT_PROVIDER" == *"github"* ]]; then
+  buildbox-run "git fetch origin \"+refs/pull/$BUILDBOX_PULL_REQUEST/merge:\""
```

Comment:
Given that we're doing `git checkout -qf \"$BUILDBOX_COMMIT\"` below (i.e. getting the referenced commit, not the `/merge` commit), this could just be `git fetch origin \"+refs/pull/$BUILDBOX_PULL_REQUEST/head:\".

The danger of trying to get the `/merge` commit is that it only works if there *is* a `/merge` commit (i.e. the PR can be merged into the destination branch without conflicts).